p2p/simulations/network.go

```diff
@@ -595,8 +622,21 @@ type Node struct {
 	// Config if the config used to created the node
 	Config *adapters.NodeConfig `json:"config"`
 
-	// Up tracks whether or not the node is running
-	Up bool `json:"up"`
+	// up tracks whether or not the node is running
+	up   bool
+	upMu sync.RWMutex
+}
+
+func (n *Node) Up() bool {
+	n.upMu.RLock()
```

Comment:
@frncmx I'm not sure why we need an `upMu` per se. I don't think that the mutex should be exclusive to `Up` field access, but rather should control access to the entire `Node` object; thus, all access/modifications to any of the `Node` fields should be using this lock. I vote to rename the mutex to be either `lock` or `nodeLock`. We shouldn't lock just on a certain field

Reply:
Note the way how `p2p/simulations.Node` embeds `adapters.Node`.

![image](https://user-images.githubusercontent.com/1735667/52869277-94c5b480-3145-11e9-9e0c-5a984ad8c12b.png)

That means a variable `node Node` provides "transparent" (e.g., `node.Addr()`) calls to methods defined by the interface and those methods also modify the state of the object state (under the interface). Those we cannot protect. So the only other field we could protect is `Node.Config`.

 We have the following methods:
```
func (n *Node) ID() enode.ID
func (n *Node) MarshalJSON() ([]byte, error)
func (n *Node) NodeInfo() *p2p.NodeInfo
func (n *Node) SetUp(up bool)
func (n *Node) String() string
func (n *Node) UnmarshalJSON(raw []byte) error
func (n *Node) Up() bool
```

From those functions only `UnmarshalJSON()` and `SetUp()` is expected to modify the state. During unmarshal I don't expect the object to be under concurrent use and that also modifies the state stored under `adapters.Node`. That we cannot protect anyway.

So from the above list we are left only with `SetUp()` and that only modifies `Node.up`. => We protect only `Node.up`.

Second thought: `Node.Config` is exported and accessed directly unfortunately.
Writes happen to the config in `Simulation.UploadSnapshot()` and `Network.snapshot()`. However, we did not get data race reports from these functions. I would try not to protect more field than it's necessary, as a rule of thumb is that the size of the critical sections should be kept minimal. That we achieve by only protecting `Node.up`. 

@justelad based on the above could you accept `upMu` as it is or would still want to extend the mutex scope? If the latter, let's have a discussion about how to protect other fields.

Reply:
@frncmx I understand that we are protecting just one field right now due to various reasons. I don't really care that the mutex is protecting just one field, I just don't think that the name of the mutex should be tied to the field's name, then once we need to protect other fields you'd need to rename the mutex anyway, or maybe introduce more locally named mutexes(no-no).

Reply:
Also, if you just need to protect one field, wouldn't `atomic.Value` be a better fit?

Reply:
@justelad renaming a mutex should not be a problem.

`atomic.Value` is nice; also preforms better. I remember I considered using it, I'm just not sure why I did not go with it. I think it was about constructing a Node. You have to be more careful with `atomic.Value` as that panics if accessed before set. So that would imply a constructor for the Node type, while `mutex` works fine without every being set.